[VTX] SmartAudio conversion to VTX table and V2.1 implementation

functionpointer opened the pull request:
This PR aims to further issue #7299 by transitioning SmartAudio to use the new VTX table facility introduced in #7251 and using it to implement SmartAudio V2.1. It should also fix #7979 and #7336.

The following features have been implemented:
- PIT mode enable and disable without power cycle
- dbm-based power selection
- persistant RACE and FREE opmodels that cause the vtx to power up in PIT mode or normal mode

At this time, vtxtables are NOT set up automatically. Therefore the user has to manually set them correctly.

For TBS Unify Pro32 Nano 5G8:

```
vtxtable bands 5
vtxtable channels 8
vtxtable band 1 BOSCAM_A  A 5865 5845 5825 5805 5785 5765 5745 5725
vtxtable band 2 BOSCAM_B  B 5733 5752 5771 5790 5809 5828 5847 5866
vtxtable band 3 BOSCAM_E  E 5705 5685 5665 5645 5885 5905 5925 5945
vtxtable band 4 FATSHARK F 5740 5760 5780 5800 5820 5840 5860 5880
vtxtable band 5 RACEBAND R 5658 5695 5732 5769 5806 5843 5880 5917
vtxtable powerlevels 3
vtxtable powervalues  14 20 26
vtxtable powerlabels 25 100 400
```

SmartAudio V1.0 devices should use the same except for the following:
```
vtxtable powerlevels 4
vtxtable powervalues 7 16 25 40
vtxtable powerlabels 25 200 500 800
```

And SmartAudio V2.0:
```
vtxtable powerlevels 4
vtxtable powervalues 0 1 2 3
vtxtable powerlabels 25 200 500 800
```

Other SmartAudio V2.1 devices should use whatever fits them. For example the TBS Unify EVO:
```
vtxtable powerlevels 4
vtxtable powervalues  14 20 26 29
vtxtable powerlabels 25 100 400 800
```
Please note that the band setup is essentially ignored by SmartAudio and it just transmits the selected band and channel indices, which are then translated by the vtx using its own frequency table. This could be changed in the future by always running the vtx in user freq mode. However, this would break channel displays of the vtx such as leds, built-in osd or CRSF telemetry.
Also note that Lua script for will be out of sync if they embed a fixed power table. This problem is also present on tramp devices (see #7540).

The code has been developed and tested with a TBS Unify Pro32 Nano 5G8 and a MATEKF722SE. Unfortunately, i don't have any other vtxes so i cannot test any other ones.
Testing especially with SmartAudio V1 and V2.0 devices would be good.

jflyper commented:
@functionpointer Is this an intermediate state for SA migration to vtxTable scheme? Do you have plan to do this?
>  always running the vtx in user freq mode

I know it involves overhauling almost everything including the "operational model" scheme.

functionpointer commented:
Maybe.

It depends on what the intention behind vtxTable is. Is the user supposed to be able to customize the behaviour of its vtx? Define custom bands? Remove unwanted power states? Or is it meant to be simply an under-the-hood system that became needed due to SA2.1 with its dynamic powerlevels?

Maybe i could create a best-of-both worlds mechanism that automatically populates vtxTable if it is empty but otherwise leaves it alone.

Docteh commented:
Is there a vtxtable command that specifies the protocol? Looks like we're very close to where an unusual VTX could have a configuration block right on the item listing.

fujin commented:
this is awesome. 💯

functionpointer commented:
> Is there a vtxtable command that specifies the protocol? Looks like we're very close to where an unusual VTX could have a configuration block right on the item listing.

The `serial` command can be used to set up the correct protocol, but it depends on the UART you soldered the vtx to. `vtx_halfduplex` needs to be set for SmartAudio, so that command could be added.

RyanHolcomb commented:
Most of this is way over my head but will this fix the SA2.1 issues I’m having after installing an Evo? When I navigate to OSD via stick commands Features > vtx SA all I see on the menu is StatX and Back. My Unify Pro with SA2.0 works fine but Evo SA 2.1 is not usable on the same quad.

functionpointer commented:
Just statx and back? That doesn't sound right. BF 4.0.2 did allow me to see and change the channel. Just all the power stuff and pitmode were broken. Maybe something else is wrong.

In any case, yes. This should get your SA2.1 device to work properly.

functionpointer commented:
So i had an idea about automatic vtxtable generation. What about an automatic or manual mode switch? In automatic mode, vtxtable would be filled automatically. This allows the vtx to be run in band/channel mode and makes setting up SmartAudio easy. Manual mode would be for advanced users and run the vtx in user freq mode to allow for maximum customization.

This would of course cost more flash as the code for both methods would need to be stored. Additionally, it would create more things that need testing.

RyanHolcomb commented:
Just installed BF4.02 Brainfpv Radix hex and SA2.1 on my Evo is still not working. It’s definitely firmware related because I can pull the vtx harness out of my Evo, plug that harness into my older Unify pro HV and SA2.0 works in my osd.

functionpointer commented on src/main/io/vtx_smartaudio.c:
```diff
@@ -27,7 +27,7 @@
 #include "platform.h"
 
-#if defined(USE_VTX_SMARTAUDIO) && defined(USE_VTX_CONTROL)
+#if defined(USE_VTX_SMARTAUDIO) && defined(USE_VTX_CONTROL) && defined(USE_VTX_TABLE)
```
The implementation currently requires `USE_VTX_TABLE`, especially for power levels. How should i handle power levels without vtxTable? Implement pulling power levels from the vtx just for this case?
Although we probably will need it for the configurator wizard anyway.

mikeller replied:
The point is that people might want to build firmware with SmartAudio but without VTX tables, for example if they do custom builds for F3. So making it possible to build 'legacy SmartAudio' without VTX table would be a plus.

functionpointer commented on src/main/cms/cms_menu_vtx_smartaudio.c:
```diff
@@ -169,17 +173,20 @@ if (saDevice.version == 2) {
-        int index = (saDevice.version == 2) ? saDevice.power : saDacToPowerIndex(saDevice.power);
-        tfp_sprintf(&saCmsStatusString[10], "%s", vtxCommonLookupPowerName(vtxCommonDevice(), index + 1));
+        saCmsPit = 1;
+        uint8_t powerindex=0;
```
Damn i must have missed those. Sorry, i am just learning the coding style.

mikeller commented:
This is starting to look good now - can you see if you can change it so that SmartAudio support can be built without VTX tables?

functionpointer commented:
Yep, im on it right now. I will push as soon as its ready.

mikeller commented on src/main/cms/cms_menu_vtx_smartaudio.c:
```diff
@@ -144,6 +146,8 @@ if (saDevice.version == 2) {
         saCmsPitFMode = 1;
     else
         saCmsPitFMode = 0;
+} else if (saDevice.version == 3){
```
There's a couple of places where missing space between `)` and `{` has snuck in again. Can you please fix this?
After that it will also need a rebase and squash into one commit.

mikeller commented on src/main/cms/cms_menu_vtx_smartaudio.c:
```diff
@@ -169,18 +173,28 @@ if (saDevice.version == 2) {
+#if defined(USE_VTX_TABLE)
+    if (vtxCommonDevice()->capability.bandCount == 0 || vtxCommonDevice()->capability.powerCount == 0) {
+        strncpy(saCmsStatusString, "PLEASE CONFIGURE VTXTABLE", sizeof(saCmsStatusString));
+        //This string is longer than 13 chars, which means once vtxtable is configured the last few characters may
```
They should by all means not linger around. If this is the case then the code does not properly zero terminate the status string when it is set in the next iteration, and this should be fixed.

functionpointer replied:
Turns out it doesn't linger because the last part of the status string is the power indication and that gets written either with `tfp_sprintf` or manually with explicit null termination. I will just remove my incorrect comment.

mikeller replied:
Even better. :-)

functionpointer commented:
Unfortunately the force-push removed all previous commits from this discussion. Commit https://github.com/betaflight/betaflight/commit/836530beacbaf577ec69963502743c25bde4b01a introduces formatting fixes. After that i rebased and squashed all of them into https://github.com/betaflight/betaflight/commit/e976ea13a74d35a3a5a82b8bf285910f30aa8c3a.

Jetrell commented:
I only recently noticed this PR after a development FW upgrade. So I came here for a read. 
I have tried to run this with a Unify Pro V2. 

With these table settings -
```
vtxtable powerlevels 4
vtxtable powervalues 0 1 2 3
vtxtable powerlabels 25 200 500 800
vtxtable bands 5
vtxtable channels 8
vtxtable band 1 BOSCAM_A A 5865 5845 5825 5805 5785 5765 5745 5725
vtxtable band 2 BOSCAM_B B 5733 5752 5771 5790 5809 5828 5847 5866
vtxtable band 3 BOSCAM_E E 5705 5685 5665 5645 5885 5905 5925 5945
vtxtable band 4 FATSHARK F 5740 5760 5780 5800 5820 5840 5860 5880
vtxtable band 5 RACEBAND R 5658 5695 5732 5769 5806 5843 5880 5917
```
I prefer to see the Band letter (R), instead of a band allocation number(5). So I entered the band tables.

What it does... is continues to transmit data on the audio channel. 
The VTX Lua script is working. But it has to keep retrying to update the changes. I assume because the VTX is continuously transmitting on the audio channel.
I don't have pit mode enabled. 

But did have `vtx_low_power_disarm = ON`  This would cause it to transmit on the audio channel all the time. But with it turned off. It would not. Until anything was changed in Lua script. Then it starts continuously transmitting on the audio channel, after a lua update to the VTX.

@functionpointer  I do understand this is still under development. But I was hoping for a few pointers, to work out whether I have missed something in the setup. And to get it working.

Thanks

functionpointer commented:
Thanks for testing @Jetrell! Looks like you are using a SmartAudio V2.0 device. I don't have any of them and thus didn't test with them. You mention continued transmissions on the audio line. So i guess you are using both audio and smartaudio at the same time? Do you hear the smartaudio transmissions in your headphones?

The continued transmissions are not intended. When SmartAudio is working properly, there should be some transmissions on power-up and some whenever settings are changed. So you are probably experiencing a bug.

So you mention this only starts once you change a setting. Which setting? power? band? channel? Does one of them work fine or does it always happen?

Jetrell commented:
@functionpointer I'm only using smart audio. But it can be heard on the DVR audio and on a small monitor I use for setup. 
So before this PR, I would only hear the smart audio transmission pulse on startup and when ever a Lua VTX change took place. 
It usually only went for 1sec or so..... Now it just keeps sending the audio pulses. And pulling over 1amp at that!

With this PR. When I had `vtx_low_power_disarm = ON`, it will continuously send the smart audio transmission from power up, continuously.
I am using Race Band CH 1 in the test today. 
But the smart audio transmission just continues regardless of what function takes place. i.e. power, band, channel.
There is a small delay when the lua change takes place. but that's all.

functionpointer commented:
Well `vtx_low_power_disarm = ON` changes vtx settings. What happens with `vtx_low_power_disarm = OFF`?

I am currently creating a custom debugging build of betaflight which will print a ton of smartaudio info when you are in cli mode. What betaflight target are you running?

Jetrell commented:
Correction on my part. Its a V3 that I'm testing. 

If `vtx_low_power_disarm = OFF`. Then it doesn't send a smart audio pulse. But as soon a setting change takes place, then it keeps transmittiing audio.
This makes sense, because `vtx_low_power_disarm = ON` will set the VTX to 25mw on power up.

When a TBS VTX is unlocked. The LED lights up red and blue together. Now it is only blue.

This target is FuryF4OSD and is driven from soft serial.
My other TBS VTX's are on different flight software, so I can't do testing with the V2.

functionpointer commented:
So you are running a TBS Unify Pro 5G8 V3?
According to https://www.team-blacksheep.com/products/prod:unify_pro, it is still using SmartAudio V2.0, but good to know. All SmartAudio versions should work and clearly it doesn't for you.
Its also very weird that the vtx gets locked!? That shouldn't happen.

Alright my debug build for FURYF4OSD [betaflight_4.1.0_FURYF4OSD.zip](https://github.com/betaflight/betaflight/files/3227368/betaflight_4.1.0_FURYF4OSD.zip)
Source code is here: https://github.com/functionpointer/betaflight/tree/sa21debug.
Flash it on your FC, make sure the settings are right (maybe do a diff all before and apply it afterwards).
Then enter the cli mode. Then power up the vtx. You should see the messages change and get more interesting than just `resending 0x1`.

That's the stuff i need. Let some messages pass and then use the `Save to File` button in the lower right corner of the betaflight configurator and post the file's contents.

Jetrell commented:
Yes. its a TBS Unify Pro 5G8 V3.

Here is the file - 
[BTFL_cli_20190528_214751.txt](https://github.com/betaflight/betaflight/files/3227430/BTFL_cli_20190528_214751.txt)

functionpointer commented:
Great job that helped me to find a bug! I forgot an index lookup for the power values.

I fixed it in this new one: [betaflight_4.1.0_FURYF4OSD.zip](https://github.com/betaflight/betaflight/files/3227579/betaflight_4.1.0_FURYF4OSD.zip)

Maybe it works perfectly now, or maybe there are more bugs. Try it and if you encounter another bug, send me the log file again. Make sure to include the messages that get created while the bug occurs. For example if it happens with the lua script, use the lua script with the cli window open.

functionpointer commented:
Oops forgot to ping @Jetrell

Jetrell commented:
Gave it a run.
It looks fine on the OSD, with the power level, channel number and band.

No continuous sending of the audio pulses. And the lua script does change all the settings, according to the OSD and TX lua screen.
But even though the VTX channel and band say they have updated. The channel and band don't change.
So I'm also wondering whether the power level is also updating... even though it says it has on the OSD and TX screen.

[BTFL_cli_20190528_224620.txt](https://github.com/betaflight/betaflight/files/3227744/BTFL_cli_20190528_224620.txt)

functionpointer commented:
Your log looks good. It appears that your vtx has powered up with 200mW and right after that betaflight sends a command to turn it down to 25mW. After that the vtx reports to be using 25mW. So that looks like it's working.

Can you post a log of you changing the channel? Maybe i can spot something.

Jetrell commented:
Sorry... I wrote an edit on my last post...   Its all fine now. Works good.
Glad you tracked it down :+1:

functionpointer commented:
Glad to hear it! One more thing: Your vtx did report to be locked. Has that resolved itself?

Jetrell commented:
Mmmm .. That would explain why it only had the blue LED on... instead of the blue and red.
It was unlocked. 
I'll have to unlock it again and get the RF power meter out to check it.

I know it was unlocked. And was before this firmware. Any chance of this some how re-locking it again?

functionpointer commented:
Really no idea. The SmartAudio implementation never locks a vtx.

You should be able to unlock it with its buttons and it should stay unlocked. You should also be able to unlock it by playing with pit mode settings. Maybe go into the osd settings (also called CMS) and set `PIT FMODE` to `PIR`. If it already shows `PIR`, just try changing it anyway. The vtx should immediately unlock.

If it doesnt...log it :D

Jetrell commented:
@functionpointer  I unlocked it again and tested it with the RF meter. All good. I'll watch it over the next while anyway. 

Thank you for your time in debugging the problem. I appreciate it.
I better let you update the PR.

doomedahab commented:
@RyanHolcomb : Where did you connect the SA wire on your SA 2.1 VTX?? To the Audio port? Because I just discovered that SA now should be soldered to the added DATA port on the VTX. I had it connected to the Audio port on my Unify Race V2 like you would on a SA 2.0 device and this caused the OSD to only show StatX, just like you describe. I swapped the cable over to the DATA port and now all is well. It would be so nice if TBS would just publish a dedicated wiring diagram for SA2.1 devices....

RyanHolcomb commented:
@doomedahab My evo is has a 7 pin harness that has audio (for mic sound) and it has a SA pin. I connect SA to the SA line. 

I can disconnect my EVO and connect my Unify Pro HV 7 pin to the exact same harness and SA2.0 works fine using the same line.

RipperDrone commented:
@functionpointer We did some testing today - looks very promising so far! :-)

Few questions:
- my friend has got a 7 bands VTX, but only 5 bands maximum can be configured in the vtxtable for selection. If he wants to run the lower bands like U, O which are bands 6 and 7 on his VTX - how can this be done with the vtxtable methodology? We tried to fool the VTX by setting up band 6 and 7 only (didn't work, selection of numbers higher than 5 is out of range) and by defining those bands as band 1 and 2 (didn't work because despite low band labels being shown, the VTX still selects bands by its internal indeces, e.g. band 1 was A and NOT O...). Any solution without increasing memory usage for everybody, mostly 5 band VTX users?
- despite your powerlevel fix not having been merged yet (into #1666 which we were using), setting a different power level by Lua DID switch the power, and all power labels were shown correctly on the OSD. What are we missing there?
- power labels are 3 digits only. For some long range VTX, 1000mW is quite common, even up to 2000mW. Is this supposed to be packed into 3 digits still, e.g. naming it 1k and 2k, respectively?

Interested to hear your view... :-)

functionpointer commented:
@RipperDrone At this time, only the indices of the selected band and channel are sent to the vtx. Additionally, they are limited to five bands with eight channels each.

This limitation could and probably should be lifted in the future.

May i ask what vtx your friend is running that supports more than eight bands? This is interesting because afaik SmartAudio was only meant to work with the fixed five bands and eight channels.

The powerlevel fix is not needed for SA2.1 devices. Only 2.0 and 1.0.

Powerlabels are user defined. 2k, 2W, MAX, or whatever else comes to mind. You could consistently use Watts and use .02 for 25mW. Or dbm values. My personal favorite is the notation sometimes used with resistors were you would use 1k2 for 1200mW.

RipperDrone commented:
@functionpointer tq for explaining, limiting labels to 3 alphanumeric characters is sensible, bought it.

My friend is not using more than 8, but 7 = more than the current limit of 5 bands. 

There are a couple of Eachines which support more bands than 5, e.g.: https://www.rcgroups.com/forums/showthread.php?3213565-Review-of-Eachine-TX806-leaf-VTX-5-8Ghz-72Ch-25mW-200mW-400mW-800mW-1000mW

RipperDrone commented:
Since all this is bogging memory onboard the FC: Does it make sense to include all channels? Or could there be a subset which ist most important to the user which he/she has to chose to put in the vtxtable...?

RipperDrone commented:
another one: XJB tx20 = 72ch

RipperDrone commented:
BoldClash has a massive 80ch claim, (but MWOSD) - seems the run for big numbers is on ;-)

Docteh commented:
At some point we'll get a VTX that just asks for a center frequency in number of mhz.

RipperDrone commented:
There's a good point in this statement, though crazy variations in number of power levels might be the next culprit. Easiest way would be to just define / make future SA revisions communicate Freq in MHz and Power in mW then?

fujin commented:
@RipperDrone SA2.1 GET/SET power is done in dBm, I'd expect that to continue going forward, FYI; that was one of the sources of trouble with the initial SA2.0-hackwards compatibility.

RipperDrone commented:
@fujin I see, this is where the strange power values come from :-)

````vtxtable powervalues  14 20 26 29

functionpointer commented:
@RipperDrone the eachine vtx looks interesting. It probably never was fully supported by betaflight. Maybe i will change that.

SA does allow the frequency to be set in MHz. I am a bit reluctant to always use that though, because it would break built-in channel display mechanisms such as leds and the built-in osd of the unify evo. Many users will probably just use the default channels and it would be great to keep the built-in display mechanisms working.

You can indeed check the SmartAudio version that is being used. Go into the CMS (configuration osd) and go into the `STATX` menu. There you will see a `STATUS` line that shows `OFFL`, `ONL V1`, `ONL V2` or `ONLV21`.

RipperDrone commented:
@functionpointer that's what we tried y'day, however we could cycle through all the settings OFF, SA1, 2.0, 2.1 with the stick. Which is the active one, ismit the one shown by default when entering the status menu entry?

RipperDrone commented:
Looking at the involved memory consumption, I'm still not 100% sure whether or not the super comfort in this PR is worth the (code size) investment for everybody. Being able to display/verify the VTX freq setting on BF OSD only but dropping the display/LED indicators might be preferrable for the flight performance  purists. IMHO letting the users chose whether they want to use barebone freq selection method only or full blown channel mapping would be the ideal solution. Just not sure how this could be implemented - looks like this would need forking the code or building + linking the SA code on the fly during the flashing procedure... Anyway, your code - your call 😃

functionpointer commented:
@RipperDrone yes i also noticed you can cycle through the different `STATUS` settings. It also changes which menu selections are available. This is especially harsh if you select `OFFL`. I am not sure why that was done this way, it is not my code. Maybe for debugging? Anyway the one shown by default is the active one.

Memory consumption can be reduced by compiling without `USE_VTX_TABLE`. In that case you get fixed bands and frequencies as well as fixed power levels for SA1.0 and 2.0. Power levels for SA2.1 devices are dynamically configured based on what the vtx reports to be capable of.

housefx commented:
Hi it´s me the 72ch vtx User ;)
The eachine vtx works just fine with SA2.0 and channel mode. 
We have tried a lot but user mode an set up a specific frequency failed as often we set one.

functionpointer commented:
@housefx interesting. Maybe the vtx does not support user frequency mode? Otherwise i would expect that to work. Betaflight limits the user freq to be between 5000 and 5999 MHz. But that is quite wide and you were probably in that range.

In this case the only way to activate the extra frequencies is to use channel mode with indices that are not supported by other vtxes.

housefx commented:
SpeedybeeF4 
Would be nice to try with indexes up to 9. 
I also have the same setup twice so could make cross testing without flashing back and forth

functionpointer commented:
Here you go.

[betaflight_4.1.0_SPEEDYBEEF4.zip](https://github.com/betaflight/betaflight/files/3242271/betaflight_4.1.0_SPEEDYBEEF4.zip)

This is a build of commit #7f4ef5f2 (as seen in PR https://github.com/betaflight/betaflight/pull/8350) with support for 9 bands.

RipperDrone commented:
9? Things are going overboard here 🤣

housefx commented:
@RipperDrone to use a 72ch vtx with 9 (9x8=72) Bands there are 9 band indices needed

housefx commented:
@functionpointer  perfect all bands works like a charm

robert-hoffmann commented:
Could this be why the lua script via my xlite is not able to change channels anymore ?
Getting out of pitmode works, but changing channels seems broken.

Suppose this needs updates on this side too ?
https://github.com/betaflight/betaflight-tx-lua-scripts/blob/master/src/SCRIPTS/BF/X7/vtx.lua

functionpointer commented:
@robert-hoffmann Yep, that would be it. PR https://github.com/betaflight/betaflight/pull/8350 will get it working again.
The lua script works in its current state. It just won't properly show any custom bands or powerlevels. Mostly cosmetic but still not ideal.

robert-hoffmann commented:
Cool, tnx !

RipperDrone commented:
@functionpointer getting a little confused about the 0-based vs. 1-based indices of buffers for powerlevels. It seems you always add one exra pseudo level to store a '---' label and such. In your example vtxtable for SA2.0, you define powerlevels as follows:

````vtxtable powerlevels 4
vtxtable powervalues 0 1 2 3
vtxtable powerlabels 25 200 500 800````

So when you want to select 25mW, the CLI command would be set vtx_power = 0, correct?

My friend @housefx has tested 1-based powervalues on his Eachine (vtxtable powervalues 1 2 3 4...), he can select 25mW with set vtx_power = 1. Are both options ok? I thought that a given VTX needs to receive the correct powervalue as a selector, and that it's always 0 based for SA2.0. Or are there VTX's with 0-based indices and others woth 1-based indices amongst the SA2.0 VTX's?

RipperDrone commented:
Have tried both options for the AKK Oscar's Backpack VTX (new version, SA2.0 compliant as per TBS definition, power levels are 25mW and 200mW):

Regardless whether I try powerlevels 0 1 or 1 2 and select the highest value in each case over CLI command 'set power_level = <x>', followed by 'save' (battery always plugged in to power VTX), VTX remains at 25mW. Yes, I have low power disarm OFF and no pitmode configured.

Running out of ideas...

robert-hoffmann commented:
Tried this this afternoon and whether i use the osd or lua (1.2.1), i still cannot change channels.

I can still get out of pit mode, so that's good ^^

This is with a tbs unify 5v ..don't remember if it's a pro or whatever, i think it's the v2. In any case it's from july 2017

functionpointer commented:
@RipperDrone Were you running the fixed (https://github.com/betaflight/betaflight/pull/8350) version? Also is you vtxTable set up properly?

The `vtx_power` CLI command does indeed start at 1. So 25mW would be 1. 0 stands for unknown power level, which is not a useful value to set `vtx_power` to.
Please note that the CLI mode disables parts of SmartAudio. If you want to change your power setting with the CLI, the setting will not be applied until you type `save` to reboot the copter. You may also note that pitmode via AUX channel and settings via OSD dont work while in CLI mode.

@Jetrell reports his SA2.0 device to be able to use all power levels with the fix in place, so it would be surprising to find something actually wrong.

functionpointer commented:
@robert-hoffmann is your vtxTable set up?

RipperDrone commented:
@functionpointer Yes, Nightly build #1669, including #8350. 

Even though vtx_power starts at 1, the vtxtable powerlevels need to be 0-based?

Yes, I always saved and rebooted, to no avail. but there are new insights: I found that the exact same VTX was working in SA on another FC. Only difference seemed to be hardwired SA, whereas it was Softserial wired on the other 2 quads that I couldn't get to work. One of them had STATX page showing bytes passed in both directions, the second one seemed dead/frozen in serial transmissions. This gave me a hint towards Softserial being the problem. I have re-soldered the SA cable to a free UART and - bingo!

Seems the old bug from last year, where Softserial has timing / pull down issues which do not seem to be fixed properly... Reference: PR https://github.com/betaflight/betaflight/pull/5076

functionpointer commented:
@RipperDrone to save flash space, vtxtable is 0-based. The conversion to 1-based and the addition of the dummy entry is done at runtime, when the vtxtable is loaded for use. The code for that is in is in [drivers/vtx_table.c](https://github.com/betaflight/betaflight/blob/master/src/main/drivers/vtx_table.c) in the function `vtxTableInit()`

Glad to hear that problem solved!

robert-hoffmann commented:
@functionpointer Totally missed this !

Just updated the tables to SA2.0. Will do a test as soon as my batteries finish charging and let you know if it's working.

I do hope BFC will have a quick drop-down for users to choose their device from that preconfigures the vtxtables ..otherwise bug reports will start overflowing with ppl complaining their SA enabled device is not working ! ^^

functionpointer commented:
@robert-hoffmann I have already added a notification in the osd menu.
It should say `PLEASE CONFIGURE VTXTABLE`. But i can see that its easy to miss.
I do agree the configurator should get a nice page for vtxtable and have some presets ready to go.

robert-hoffmann commented:
@functionpointer yep, with updated vtxtable it's working just fine, via osd & lua 
tnx 👍

Docteh commented:
@functionpointer the wiki would be a good start for vtxtable presets. (disclaimer: have not checked yet)

wildflyer commented:
hi,
how can I replace the vtx table ? in cli ?

RipperDrone commented:
Yes, commands vtxtable ... See comments on top

wildflyer commented:
I tried in bf 3.5 and in bf 4.0.3, there is no command in cli named "vtxtable".

RipperDrone commented:
correct, only available in 4.1 nightly code versions. Pull from jenkins.

wildflyer commented:
where can I download and how to flash it ? I have a nox board.

amalberg commented:
I've tried using same table for the unify pro 32 for my unify race pro hv v2 (also 25,100,400) and it was still not matching up. I flashed to night build 1680. Cannot get the power or channels to correspond correctly? At least it's working so I will be able to fly now

sergeymishin commented:
> where can I download and how to flash it ? I have a nox board.
> 
> edit: I found the hex files and I will try it, thx for help! :-)
> 
> edit2:
> ok, everything working now, thx.

Where did you get one?
I have TBS UNIFY PRO HV RACE 2, and it doesn't work with 4.0.3 on my MATEKF405. Please help!

functionpointer commented:
Ill just quote RipperDrone

> correct, only available in master/4.1 nightly code versions. Pull from jenkins.

sergeymishin commented:
> Ill just quote RipperDrone
> 
> > correct, only available in master/4.1 nightly code versions. Pull from jenkins.

Unfortunately, I don't know how to pull from jenkins. Maybe there are some already compiled hex files somewhere?

functionpointer commented:
https://ci.betaflight.tech/view/Firmware/job/Betaflight/lastBuild/artifact/obj/

amalberg commented:
Have a question about the power levels. 
On the unify race there's 3 power levels and pit mode. Does that mean I need to set 4 or is there a different trigger to get into pit mode? 
It seemed as though when I was on 25 setting it wad in pit mode then 100 on the screen was 25mw. This was checked with an rx explorer During a race event.

functionpointer commented:
@amalberg pit mode is separate from power levels. You can be in 100mW and pit mode at the same time. That would mean the vtx is at very low power, but as soon as you leave pit mode it goes up to 100mW.

The osd menu (CMS) clearly shows this with two separate settings for power and pit mode.
Pit mode can also be enabled and disabled with an aux channel.

The vtxtable does not have any extra entries for pitmode. This is also shown right at the top of this page in the example vtxtable setups.

amalberg commented:
Yeah that pit mode setting was disabled but I still had low power according to the rf explorer. Then when I went to 100 it was the same as everyone else.

functionpointer commented:
@amalberg what are the LEDs on the vtx showing?
blue? blue+red? or red?
Just red means pit mode. Just double checking if the problem is in betaflight or on the vtx.

amalberg commented:
Ahh ok will check it out next time I have a fly. From memory it was both but not sure. Its already a replaced vtx due to first one not having the correct channel map internally.

RipperDrone commented:
Two space characters in Unify Pro32 Nano powerlevels vtxtable (in front of the 14) are ok/intentional? Just verifying if people use this table for copy/paste...

vtxtable powervalues  14 20 26

Looks like command parser doesn't care about number of space separators btw data? 🙂

functionpointer commented:
@RipperDrone yes, extra spaces are ignored. In fact, when you type `vtxtable` into the cli you will find betaflight itself adds extra spaces for column alignment.

housefx commented:
Tried today the newest nightly with 8 bands implementation. 
Only the standard 5 bands worked over SA
Flashed back to the customized hex for my FC with the 9 Bands and everything works again. 
Tried also with the exactly same config (diff all) with vtxtables set to 8Bands. 
Custom hex =perfekt 
Nightly = only the normal bands work. 

Any idea why

functionpointer commented:
@housefx there is a check in io/smartaudio.c that only allows five bands. I have removed it for the custom build.

The check will be gone from master once #8380 is merged. That should allow you to run normal nightlies with eight bands. Nine bands will still require a custom build with `VTX_TABLE_MAX_BANDS` in drivers/vtx_table.h set to 9.

housefx commented:
@functionpointer ok that’s the cause. 
Thx for the quick reply and answer. 
8 or 9 bands are both fine for me, want to use Band 7 & 8 so all perfect in the future. 
Will check with the #8380 and later nightly once they pop out.
